crawl/crawl.go

```diff
@@ -1,178 +1,218 @@
+		rate := time.Duration(float64(e.index) / float64(elapsed))
+		var eta time.Duration
+		if rate > 0 {
+			eta = time.Duration(endIndex-e.index) / rate * time.Second
 		}
-	}
-
-	// Start a worker which reads entries from the fetchers. If it receives an
-	// entry for which it has not already processed the immediately preceding
-	// entry, it buffers it and waits to process it until the intervening entries
-	// have been backfilled. When processing an entry, it compares it to the
-	// immediately prior entry: if the timestamp has traveled backwards in time,
-	// it outputs the *prior* entry, on the assumption that its timestamp was
-	// incorrectly forward in time.
-	buffer := NewHeap[entryData]()
-	nextIndex := *startIndex
-	last := entryData{index: -1}
-	var wg sync.WaitGroup
-	wg.Add(1)
-	go func() {
-		defer wg.Done()
-		for e := range entries {
-			// If it's not the entry we want, save it for later.
-			if e.index != nextIndex {
-				buffer.Push(e)
-				continue
-			}
-
-			// If this is the first entry we're looking for, initialize our last-seen
-			// tracker so we have a basis for comparison.
-			if last.index == -1 {
-				last = e
-				nextIndex = e.index + 1
-				continue
-			}
 
-			// This is the next entry we were looking for. Process it.
-			processEntry(last, e)
-			last = e
-			nextIndex += 1
-			if nextIndex%1000 == 0 {
-				fmt.Printf("Processed up to index %d\n", nextIndex)
-			}
+		if e.index%98689 == 0 {
+			log.Printf("index %d of %d; %2.1f%%; ETA %s; leaf %s; root %s",
+				e.index, endIndex, fraction*100, eta,
+				b64(e.merkleLeafHash), b64(rootHash))
 
-			// Try to process the buffer, just in case we've caught up to it.
-			for buffer.Len() > 0 && buffer.Peek().index == nextIndex {
-				e = buffer.Pop()
-				processEntry(last, e)
-				last = e
-				nextIndex += 1
-				if nextIndex%1000 == 0 {
-					fmt.Printf("Processed up to index %d\n", nextIndex)
+			go func(e entryData, rootHash []byte) {
+				err := verify(logClient, e, rootHash)
+				if err != nil {
+					log.Fatal(err)
 				}
-			}
+			}(e, rootHash)
 		}
-	}()
-
-	// Set up a context and a signal-catcher to cancel the context so we can
-	// break out cleanly if we need to.
-	ctx, cancel := context.WithCancel(context.Background())
-	go func() {
-		sigChan := make(chan os.Signal, 1)
-		signal.Notify(sigChan, syscall.SIGTERM)
-		signal.Notify(sigChan, syscall.SIGINT)
-		signal.Notify(sigChan, syscall.SIGHUP)
-
-		<-sigChan
-		cancel()
-
+	}
+	if e.index != endIndex-1 {
+		log.Printf("channel closed early? processed %d entries; expected %d", e.index, endIndex)
+	}
+	log.Printf("Final entry %d. Elapsed %s. Root hash %s",
+		e.index, time.Since(start), b64(rootHash))
+	if !bytes.Equal(rootHash, sth.SHA256RootHash[:]) {
+		log.Fatalf(
+			"calculated root hash differed from log's reported root hash at tree size %d: calculated %s, log reported %s",
+			sth.TreeSize,
+			b64(rootHash),
+			b64(sth.SHA256RootHash[:]))
+	} else {
+		log.Fatalf(
```

Comment:
I think you meant `log.Printf` so the `os.Exit(0)` actually gets to run.
```suggestion
		log.Printf(
```